pkg/operator/ceph/cluster/mon/mon_test.go

```diff
@@ -278,24 +280,210 @@ func TestPersistMons(t *testing.T) {
+	assertEndpointSlicePorts(t, ep.Ports, expectedPorts)
+	assert.Equal(t, map[string]string{"key": "value"}, cm.Annotations)
+
 	// Persist mon b, and remove mon a for simply testing the configmap is updated
 	c.ClusterInfo.InternalMonitors["b"] = &cephclient.MonInfo{Name: "b", Endpoint: "4.5.6.7:3300"}
 	delete(c.ClusterInfo.InternalMonitors, "a")
-	err = c.persistExpectedMonDaemons()
+	err = c.persistExpectedMonDaemonsInConfigMap()
+	assert.NoError(t, err)
+	err = c.persistExpectedMonDaemonsAsEndpointSlice()
 	assert.NoError(t, err)
 
 	cm, err = c.context.Clientset.CoreV1().ConfigMaps(c.Namespace).Get(context.TODO(), EndpointConfigMapName, metav1.GetOptions{})
 	assert.NoError(t, err)
 	assert.Equal(t, "b=4.5.6.7:3300", cm.Data[EndpointDataKey])
+
+	ep, err = c.context.Clientset.DiscoveryV1().EndpointSlices(c.Namespace).Get(context.TODO(), endpointSliceNameIPv4, metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.Equal(t, "4.5.6.7", ep.Endpoints[0].Addresses[0])
+	assert.Len(t, ep.Ports, len(expectedPorts))
+	assertEndpointSlicePorts(t, ep.Ports, expectedPorts)
+	assert.Equal(t, map[string]string{"key": "value"}, cm.Annotations)
+}
+
+func TestCreateEndpointSliceForAddresses(t *testing.T) {
+	clientset := test.New(t, 1)
+	ownerInfo := cephclient.NewMinimumOwnerInfoWithOwnerRef()
+	c := New(context.TODO(), &clusterd.Context{Clientset: clientset}, "ns", cephv1.ClusterSpec{}, ownerInfo)
+
+	ipv4Addresses := []string{"10.0.0.1", "10.0.0.2", "10.0.0.3"}
+	ipv4Mons := []cephclient.MonInfo{
+		{Name: "a", Endpoint: fmt.Sprintf("%s:6789", ipv4Addresses[0])},
+		{Name: "b", Endpoint: fmt.Sprintf("%s:6789", ipv4Addresses[1])},
+		{Name: "c", Endpoint: fmt.Sprintf("%s:6789", ipv4Addresses[2])},
+	}
+	ipv6Addresses := []string{"2001:db8::1", "2001:db8::2", "2001:db8::3"}
+	ipv6Mons := []cephclient.MonInfo{
+		{Name: "d", Endpoint: fmt.Sprintf("[%s]:6789", ipv6Addresses[0])},
+		{Name: "e", Endpoint: fmt.Sprintf("[%s]:6789", ipv6Addresses[1])},
+		{Name: "f", Endpoint: fmt.Sprintf("[%s]:6789", ipv6Addresses[2])},
+	}
+	expectedPortsNoRequireMsgr2 := []struct {
+		Name string
+		Port int32
+	}{
+		{DefaultMsgr2PortName, DefaultMsgr2Port},
+		{DefaultMsgr1PortName, DefaultMsgr1Port},
+	}
+
+	// IPv4 test
+	c.ClusterInfo.InternalMonitors = map[string]*cephclient.MonInfo{}
+	for _, mon := range ipv4Mons {
+		c.ClusterInfo.InternalMonitors[mon.Name] = &mon
+	}
+
+	err := c.persistExpectedMonDaemonsAsEndpointSlice()
+	assert.NoError(t, err)
+
+	epSliceIPv4, err := c.context.Clientset.DiscoveryV1().EndpointSlices(c.Namespace).Get(context.TODO(), endpointSliceNameIPv4, metav1.GetOptions{})
+	assert.NoError(t, err)
+
+	assert.Equal(t, discoveryv1.AddressTypeIPv4, epSliceIPv4.AddressType)
+	assert.Len(t, epSliceIPv4.Endpoints, 1)
+	assert.ElementsMatch(t, ipv4Addresses, epSliceIPv4.Endpoints[0].Addresses)
+	assert.Len(t, epSliceIPv4.Ports, len(expectedPortsNoRequireMsgr2))
+	assertEndpointSlicePorts(t, epSliceIPv4.Ports, expectedPortsNoRequireMsgr2)
```

Comment:
Why check for the length here, if we check the whole array anyway? `assertEndpointSlicePorts()` already contains a length check.

Anyway, I suggest to simply use `assert.ElementsMatch()` and get rid of the `assertEndpointSlicePorts()` method, like so: https://github.com/haslersn/rook/commit/7be0673075863ea915fd9a690246eb4acdcca4ab